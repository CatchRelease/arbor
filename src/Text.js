import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { color, fontSize, space, textAlign } from 'styled-system';

const baseStyles = ({ theme }) => css`
  font-family: ${theme.brandFont};
  line-height: ${theme.lineHeights.large};
`;

const Text = styled.p`
  ${color};
  ${fontSize};
  ${space};
  ${textAlign};
  ${baseStyles};
`;

Text.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes
};

Text.p = Text.withComponent('p');
Text.span = Text.withComponent('span');

export default Text;
