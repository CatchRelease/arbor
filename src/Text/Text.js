import React from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import {
  color,
  fontSize,
  fontWeight,
  lineHeight,
  maxWidth,
  minWidth,
  overflow,
  space,
  textAlign
} from 'styled-system';

import textTransform from '../utils/textTransform';
import textOverflow from '../utils/textOverflow';
import whiteSpace from '../utils/whiteSpace';

const baseStyles = ({ theme }) => css`
  font-family: ${theme.brandFont};
`;

const Text = styled.p`
  ${baseStyles};

  ${color};
  ${fontSize};
  ${fontWeight};
  ${lineHeight};
  ${maxWidth};
  ${minWidth};
  ${overflow};
  ${space};
  ${textAlign};
  ${textTransform};
  ${textOverflow};
  ${whiteSpace};
`;

Text.propTypes = {
  ...color.propTypes,
  ...fontSize.propTypes,
  ...fontWeight.propTypes,
  ...lineHeight.propTypes,
  ...maxWidth.propTypes,
  ...minWidth.propTypes,
  ...overflow.propTypes,
  ...space.propTypes,
  ...textAlign.propTypes,
  ...textTransform.propTypes,
  ...textOverflow.propTypes,
  ...whiteSpace.propTypes
};

Text.defaultProps = {
  color: 'text.default',
  fontSize: 'size4',
  lineHeight: 'small'
};

// Deprecated -- use <Text as="span"></Text>
Text.Span = (props) => <Text as="span" {...props} />;

export default Text;
