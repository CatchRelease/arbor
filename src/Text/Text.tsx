import { HTMLAttributes } from 'react';
import { css, Theme } from '@emotion/react';
import styled from '@emotion/styled';
import {
  color,
  ColorProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  lineHeight,
  LineHeightProps,
  maxWidth,
  MaxWidthProps,
  minWidth,
  MinWidthProps,
  overflow,
  OverflowProps,
  space,
  SpaceProps,
  textAlign,
  TextAlignProps
} from 'styled-system';

import { WithColorPropFix } from '../colorPropFix';
import textDecoration, { TextDecorationProps } from '../utils/textDecoration';
import textOverflow, { TextOverflowProps } from '../utils/textOverflow';
import textTransform, { TextTransformProps } from '../utils/textTransform';
import whiteSpace, { WhitespaceProps } from '../utils/whiteSpace';

const baseStyles = ({ theme }: { theme: Theme }) => css`
  font-family: ${theme.brandFont};
`;

export type Props = Omit<HTMLAttributes<HTMLElement>, 'color'> &
  ColorProps &
  FontSizeProps &
  FontWeightProps &
  LineHeightProps &
  MaxWidthProps &
  MinWidthProps &
  OverflowProps &
  SpaceProps &
  TextAlignProps &
  TextDecorationProps &
  TextTransformProps &
  TextOverflowProps &
  WhitespaceProps;

const Text = styled.p<Props>`
  ${baseStyles};

  ${color};
  ${fontSize};
  ${fontWeight};
  ${lineHeight};
  ${maxWidth};
  ${minWidth};
  ${overflow};
  ${space};
  ${textAlign};
  ${textDecoration};
  ${textTransform};
  ${textOverflow};
  ${whiteSpace};
`;

Text.defaultProps = {
  color: 'text.default',
  fontSize: 'size4',
  lineHeight: 'small'
};

const TextWithColorPropFix: WithColorPropFix<typeof Text> = Text;

export default TextWithColorPropFix;
