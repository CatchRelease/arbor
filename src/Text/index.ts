export { default, Props as TextProps } from './Text';
