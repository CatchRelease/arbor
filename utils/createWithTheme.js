import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
