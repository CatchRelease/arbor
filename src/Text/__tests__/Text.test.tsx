import createWithTheme from '../../../utils/createWithTheme';
import Text, { Props as TextProps } from '../Text';

describe('<Text />', () => {
  it('properly renders a Text component', () => {
    const tree = createWithTheme(
      <Text>The quick brown fox jumps over the lazy dog</Text>
    );

    expect(tree).toMatchSnapshot();
  });

  it('supports textDecoration prop', () => {
    const tree = createWithTheme(
      <Text textDecoration="line-through">
        The quick brown fox jumps over the lazy dog
      </Text>
    );

    expect(tree).toMatchSnapshot();
  });

  it('exports props compatible with itself', () => {
    // This verifies that our Text component and TextProps are compatible after
    // all our color-prop shenanigans.
    const props: TextProps = { color: 'gray' };

    const tree = createWithTheme(
      <Text {...props}>The quick brown fox jumps over the lazy dog</Text>
    );

    expect(tree).toMatchSnapshot();
  });
});
