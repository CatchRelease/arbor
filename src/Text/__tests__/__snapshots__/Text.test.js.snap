// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> properly renders a Text component 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #25272D;
}

<p
  className="emotion-0 emotion-1"
  color="text.default"
>
  The quick brown fox jumps over the lazy dog
</p>
`;
